docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
+  pods to query for metrics within their own namespace.
+
+- separating querying from pushing: we should allow certain users to push, and
+  certain users to query, but not necessarily tie the two together
+
+- allowing different push scope permissions: we should be able to specify that
+  some users are only allowed to push metrics for the pods within their own
+  RC or namespace, while others should be allowed to push for any object in
+  the cluster.
+
+Proposed Solution
+-----------------
+
+Like in Kubernetes, authentication and authorization should be split into
+separate interfaces.  Each would be configured using a URL flag at the command
```

Comment:
As in placing the configuration in a separate file?  While I'm not necessarily against that, I suspect that most of the auth mechanisms would work fine with just a "URL" configuration (note, this doesn't necessarily mean that the authentication or authorization is based on a remote HTTP call -- Heapster just uses a quasi-URL scheme for specifying configuration of sinks and sources, so this would probably work fine here as well).